### PR TITLE
[CIVIS-2325] DEP upgrade to Python 3.12, update dependency versions

### PR DESCRIPTION
**Pull request by jacksonlee-civis**
This pull request upgrades the ds-py image to Python 3.12 and updates the dependencies' versions (notably civis-python v2.0.0!).

---

- [x] (For Civis employees only) Reference to a relevant ticket in the pull request title
- [x] Changelog entry added to `CHANGELOG.md` at the repo's root level
- [x] Description of change in the pull request description
- [x] The CircleCI builds have all passed

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,29 +1,17 @@
...
           command: docker run ds-python /bin/bash -c "echo BUILDS OK"
       - run:
-          name: Check that scipy links to OpenBLAS
-          command: docker run ds-python python -c "from scipy.linalg import _fblas"
-      - run:
-          name: Check that numpy imports from bash shell
-          command: docker run -t ds-python /bin/bash -c "python -c 'import numpy'"
-      - run:
-          name: Verify that numpy does not link to MKL
-          command: docker run ds-python python -c "from numpy.distutils import system_info; assert system_info.get_info('mkl') == {}"
```
Comment (jacksonlee-civis) [inline .circleci/config.yml:23]: `numpy.distutils` has been removed for Python 3.12. This check was added likely for the previous conda set-up, but we aren't using conda anymore.

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,29 +1,17 @@
...
       - run:
           name: Build container
-          command: docker build -t ds-python .
+          command: docker build --target test -t ds-python .
       - run:
           name: Verify build completed
           command: docker run ds-python /bin/bash -c "echo BUILDS OK"
       - run:
-          name: Check that scipy links to OpenBLAS
-          command: docker run ds-python python -c "from scipy.linalg import _fblas"
```
Comment (jacksonlee-civis) [inline .circleci/config.yml:17]: Moving this and other tests to the python unitttest script I'm adding below for readability. In general, as long as we verify that the key packages can be imported, we should be good.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,29 +1,17 @@
...
-          command: docker run -t ds-python /bin/bash -c "python -c 'import numpy'"
-      - run:
-          name: Verify that numpy does not link to MKL
-          command: docker run ds-python python -c "from numpy.distutils import system_info; assert system_info.get_info('mkl') == {}"
-      - run:
-          name: Validate compatability of civis-python with other packages
-          command: docker run ds-python /bin/bash -c "python -c 'import civis'"
-      - run:
-          name: Run numpy unit tests
-          command: docker run ds-python /bin/bash -c 'pip install pytest hypothesis; python -c "import numpy; numpy.test()"'
```
Comment (jacksonlee-civis) [inline .circleci/config.yml:29]: Another numpy check that was likely added due to conda. Not ported in the new python script below (which saves a lot of time in the CircleCI buid!).

**Comment by jacksonlee-civis** [comment]
@thatguyinabeanie Mind reviewing this pull request? This is not urgent, so next week would be fine too.

File: README.md [hunk 4]
```diff
@@ -68,10 +66,6 @@ default location for staging temporary files to the /tmp directory.
 The normal default is /shm. /shm is a RAM disk which defaults to a 64 MB size
 in Docker containers, too small for typical scientific computing.
 
-# Install a new PyPi Package Dependency
-1. Add new dependency to `requirements-core.txt` file. Include the version
-2. Run script `generate-requirements-full.sh`
```
Comment (jacksonlee-civis) [inline README.md:73]: Dropping this bit to discourage adding new packages in general. The more packages we consider as "core", the harder it is for us to maintain a resolvable lock / requirements-full.txt.

File: Dockerfile [hunk 5]
```diff
@@ -35,7 +35,14 @@ RUN pip install -r requirements-full.txt && \
...
-ENV VERSION=7.0.0 \
+ENV VERSION=7.1.0 \
   VERSION_MAJOR=7 \
-  VERSION_MINOR=0 \
+  VERSION_MINOR=1 \
   VERSION_MICRO=0
+
+FROM production AS test
+COPY .circleci/test_image.py .
+COPY CHANGELOG.md .
```
Comment (thatguyinabeanie) [inline Dockerfile:45]: why is the changelog.md file needed?

**Review by thatguyinabeanie — CHANGES_REQUESTED** [review]
1 question. otherwise lgtm

File: Dockerfile (hunk 5 above)
Comment (jacksonlee-civis) [reply Dockerfile:45]: `CHANGELOG.md` is for the test that verifies if the latest version number in the changelog matches the environment variable `VERSION` in the image. It's too easy for us to forget updating one of these things.

Comment (thatguyinabeanie) [reply Dockerfile:45]: nice!! i like this. im gunna have to steal this for some of our repos

**Review by thatguyinabeanie — APPROVED** [review]
LGTM

**Pull request merged**